File: backend/routing_rules.py
import fnmatch
from dataclasses import dataclass


@dataclass(frozen=True)
class RoutingRule:
    file_pattern: str
    reviewer_role: str
    auto_approve: bool


DEFAULT_ROUTING_RULES = [
    RoutingRule("schema.prisma", "Lead Backend Engineer", False),
    RoutingRule("*.prisma", "Lead Backend Engineer", False),
    RoutingRule("*.sql", "Database Engineer", False),
    RoutingRule("*.py", "Backend Engineer", False),
    RoutingRule("*.ts", "Backend Engineer", False),
    RoutingRule("*.tsx", "Frontend Engineer", False),
    RoutingRule("*.js", "Frontend Engineer", False),
    RoutingRule("*.css", "UI/UX Lead", True),
    RoutingRule("*.md", "Docs Reviewer", True),
    RoutingRule("*.txt", "Docs Reviewer", True),
]

ROUTING_RULES = DEFAULT_ROUTING_RULES

DOC_EXTENSIONS = (".md", ".txt", ".rst")


def is_docs_file(filename: str) -> bool:
    name = filename.lower()
    if name.endswith(DOC_EXTENSIONS):
        return True
    if name.startswith("docs/") or "/docs/" in name:
        return True
    return any(token in name for token in ("readme", "changelog", "contributing"))


def route_files(
    files: list[str], rules: list[RoutingRule] | None = None
) -> tuple[list[str], bool]:
    """Return (reviewer roles, docs_only) for the touched files."""
    active_rules = rules if rules is not None and len(rules) > 0 else DEFAULT_ROUTING_RULES
    reviewers: list[str] = []
    docs_only = bool(files)
    for filename in files:
        docs_only = docs_only and is_docs_file(filename)
        for rule in active_rules:
            if fnmatch.fnmatch(filename, rule.file_pattern) and rule.reviewer_role not in reviewers:
                reviewers.append(rule.reviewer_role)
    if not reviewers:
        reviewers.append("Backend Engineer")
    return reviewers, docs_only

File: frontend/convex/routingRules.ts
import { v } from 'convex/values'
import { mutation, query } from './_generated/server'

export const DEFAULT_RULES = [
  { file_pattern: 'schema.prisma', reviewer_role: 'Lead Backend Engineer', auto_approve: false },
  { file_pattern: '*.prisma', reviewer_role: 'Lead Backend Engineer', auto_approve: false },
  { file_pattern: '*.sql', reviewer_role: 'Database Engineer', auto_approve: false },
  { file_pattern: '*.py', reviewer_role: 'Backend Engineer', auto_approve: false },
  { file_pattern: '*.ts', reviewer_role: 'Backend Engineer', auto_approve: false },
  { file_pattern: '*.tsx', reviewer_role: 'Frontend Engineer', auto_approve: false },
  { file_pattern: '*.js', reviewer_role: 'Frontend Engineer', auto_approve: false },
  { file_pattern: '*.css', reviewer_role: 'UI/UX Lead', auto_approve: true },
  { file_pattern: '*.md', reviewer_role: 'Docs Reviewer', auto_approve: true },
  { file_pattern: '*.txt', reviewer_role: 'Docs Reviewer', auto_approve: true },
]

export const getRoutingRules = query({
  args: {},
  handler: async (ctx) => {
    return await ctx.db.query('routing_rules').collect()
  },
})

export const createRoutingRule = mutation({
  args: {
    file_pattern: v.string(),
    reviewer_role: v.string(),
    auto_approve: v.boolean(),
  },
  handler: async (ctx, args) => {
    return await ctx.db.insert('routing_rules', {
      file_pattern: args.file_pattern,
      reviewer_role: args.reviewer_role,
      auto_approve: args.auto_approve,
    })
  },
})

export const updateRoutingRule = mutation({
  args: {
    ruleId: v.id('routing_rules'),
    file_pattern: v.string(),
    reviewer_role: v.string(),
    auto_approve: v.boolean(),
  },
  handler: async (ctx, args) => {
    await ctx.db.patch('routing_rules', args.ruleId, {
      file_pattern: args.file_pattern,
      reviewer_role: args.reviewer_role,
      auto_approve: args.auto_approve,
    })
  },
})

export const deleteRoutingRule = mutation({
  args: { ruleId: v.id('routing_rules') },
  handler: async (ctx, args) => {
    await ctx.db.delete('routing_rules', args.ruleId)
  },
})

export const seedDefaultRoutingRules = mutation({
  args: {},
  handler: async (ctx) => {
    const existing = await ctx.db.query('routing_rules').first()
    if (existing) return { seeded: false }
    for (const rule of DEFAULT_RULES) {
      await ctx.db.insert('routing_rules', rule)
    }
    return { seeded: true, count: DEFAULT_RULES.length }
  },
})
